src/WebJobs.Extensions.DurableTask/ContextImplementations/DurableOrchestrationContext.cs

```diff
@@ -275,10 +279,98 @@ Task IInterleavingContext.CallActorAsync(ActorId actorId, string operationName,
         }
 
         /// <inheritdoc/>
-        Task<IDisposable> IInterleavingContext.LockAsync(params ActorId[] actors)
+        async Task<IDisposable> IInterleavingContext.LockAsync(params ActorId[] actors)
         {
             this.ThrowIfInvalidAccess();
-            throw new NotImplementedException(); // TODO
+            if (this.ContextLocks != null)
+            {
+                throw new LockingRulesViolationException("Cannot acquire more locks when already holding some locks.");
+            }
+
+            if (actors == null || actors.Length == 0)
+            {
+                throw new ArgumentException("The list of actors to lock must not be null or empty.", nameof(actors));
+            }
+
+            // acquire the locks in a globally fixed order to avoid deadlocks
+            Array.Sort(actors);
+
+            // remove duplicates if necessary. Probably quite rare, so no need to optimize more.
+            for (int i = 0; i < actors.Length - 1; i++)
+            {
+                if (actors[i].Equals(actors[i + 1]))
+                {
+                    actors = actors.Distinct().ToArray();
+                    break;
+                }
+            }
+
+            // use a deterministically replayable unique ID for this lock request, and to receive the response
+            var lockRequestId = this.NewGuid();
+
+            // send lock request to first actor in the lock set, who will forward it to the next, etc.
+            var target = new OrchestrationInstance() { InstanceId = ActorId.GetSchedulerIdFromActorId(actors[0]) };
+            var request = new RequestMessage()
+            {
+                Id = lockRequestId,
+                ParentInstanceId = this.InstanceId,
+                LockSet = actors,
+                Position = 0,
+            };
+
+            this.LockRequestId = lockRequestId.ToString();
+
+            var jrequest = JToken.FromObject(request, MessagePayloadDataConverter.DefaultSerializer);
+            this.InnerContext.SendEvent(target, "op", jrequest);
```

Comment:
It's a hack to avoid typenames being added to the JSON. I really don't want the typenames in there, but DTFx does not give me access to the JSON serializer settings so I can't control the `TypeNameHandling` setting. Maybe we should revise DTFx so it accepts custom serializer settings.